lite/micro: Fix bug in tensor lifetime calculation.

suphoff opened the pull request:
Fix for issue #35121

The tensor lifetime may be incorrectly calculated in MicroAllocator::FinishTensorAllocation() if the same sensor is used multiple times as inputs to different operations or is used as input/output or variable of the graph.

petewarden reviewed (changes requested):
Thanks for the fix! To make sure we don't regress, could you see if it's possible to add a test that fails before this change in micro_allocator_test.cc? I took a quick look, and it may be a bit complex since you'll probably need to create a new tflite::testing::GetMockModel() function that triggers the error case. I'm happy to collaborate on the test though, if it isn't straightforward.

suphoff commented:
@petewarden : Looks easy enough to add more nodes to the MockModel and adjust existing tests that use it. 

I will open a new PR for the test code as I assume we want to observe if failing before this PR is pulled.

hf-sky-hf commented:
Until someone comes to talk to me in person we done I have no idea who
Google and you are up too

On Wed, Dec 18, 2019, 19:15 Stephan Uphoff <notifications@github.com> wrote:

> @petewarden <https://github.com/petewarden> : Looks easy enough to add
> more nodes to the MockModel and adjust existing tests that use it.
>
> I will open a new PR for the test code as I assume we want to observe if
> failing before this PR is pulled.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/35123?email_source=notifications&email_token=AOBPXOQ6MVOY2XCTOP4ZM23QZK4KFA5CNFSM4J25DPH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHH5PDY#issuecomment-567269263>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOBPXOT23VYHHC3S2EVRMWDQZK4KFANCNFSM4J25DPHQ>
> .
>

suphoff committed:
lite/micro: Fix bug in tensor lifetime calculation.

Fix for issue #35121

Testing:
Add a second operation to the MockModel sharing inputs with the first
operation to catch #35121 (Tensor lifetime incorrectly calculated on
multiple use) and verify the fix.

suphoff commented:
@petewarden : Tests are in #35294

gbaned commented:
@suphoff  Can you please resolve conflicts? Thanks!

suphoff commented:
@gbaned : Conflicts removed - conflicting changes only addressed a subset of the issues.

gbaned commented:
@suphoff  Can you please resolve conflicts? Thanks!

njeffrie reviewed (changes requested):
Whoops - looks like we still need a test or two to catch this condition.  Sorry for the premature approval.

suphoff commented:
@njeffrie : Test is in #35294 (and should fail without this bug fix )

suphoff commented:
Merged the allocation tests from #35294  into this pull request as requested by @petewarden
